### ISSUE TITLE
how to take 2d tensor as input?

### ISSUE THREAD

**Issue by huiyiqun** [Dobiasd/frugally-deep#34]
I have a sequence of several features, says (number of samples in the sequence) * (number of features in a sample) points.

In Keras, I could describe it as a tensor with a shape of (N, 3) (N is the number of samples in the sequence while 3 is the number of features in a sample).

So:

1. What's the correct shape3 to use for fdeep?
2. How should I convert a vector of floats into the input of `predict`?

I have searched the document and found nothing about using 2d tensor as the input of `predict`

**Comment by huiyiqun** [comment]
I have tried following code

```
        model.predict_class({model.generate_dummy_inputs()});
```

and failed

```
Parsing json ... done. elapsed time: 0.008958 s
Running test 1 of 1 ... done. elapsed time: 0.000603 s
terminate called after throwing an instance of 'std::runtime_error'
  what():  test failed: output=0 pos=0,0,0 value=-nan target=8.3679e-18

```

**Comment by huiyiqun** [comment]
I've found that the problems is the verify-related code.(pls fix it)

Following code works:

```
const auto model = fdeep::load_model("model.json", false);
mode.predict_class({model.generate_dummy_inputs()});
```

And the shape of the input could be retrieved by:

```
const shape = model.get_input_shapes()[0];
std::cout << shape.depth_ << '\t' << shape.height_ << '\t' << shape.width_ << std::endl;
```

**Comment by Dobiasd** [comment]
Interesting, thanks for the report. That should not happen. ;)

Could you please provide a minimal python code to create a model that reproduces this error?

**Comment by huiyiqun** [comment]
Following is simplified code without the part about the dataset:

```python
import numpy as np

from keras.models import Sequential
from keras.layers import Dense, Flatten, Conv1D, MaxPooling1D, Activation


def main():
    N = 50
    model = Sequential([
            Conv1D(16, input_shape=(N, 3), kernel_size=3, padding='same'),
            Activation('relu'),
            MaxPooling1D(3, padding='same'),
            Flatten(),
            Dense(64),
            Activation('relu'),
            Dense(64),
            Activation('relu'),
            Dense(3),
            Activation('softmax'),
    ])
    model.compile(optimizer='rmsprop', loss='binary_crossentropy')
    model.fit(np.random.rand(10, N, 3), np.random.rand(10, 3), epochs=10)
    model.save('test_model.h5', include_optimizer=False)


if __name__ == '__main__':
    main()
```

It generate [test_model.txt](https://github.com/Dobiasd/frugally-deep/files/1841487/test_model.txt), which works correctly.

The only difference with my complete codebase is the dataset, which generate [true_model.txt](https://github.com/Dobiasd/frugally-deep/files/1841488/true_model.txt).

Maybe you could give the model a try.

**Comment by Dobiasd** [comment]
OK, I was able to reproduce your error. `test_model` works, `true_model` throws an exception during validation. The two JSON files look (disregarding the weight) exactly the same. I'll investigate.

**Comment by Dobiasd** [comment]
In case you need a quick workaround:

If you use
```cpp
#define FDEEP_FLOAT_TYPE double
#include "fdeep/fdeep.hpp"
```
`true_model` also works.

**Comment by Dobiasd** [comment]
By using `#include <fenv.h>` and `feenableexcept(FE_ALL_EXCEPT & ~FE_INEXACT);` one lands in `softmax_layer`, where at some point `std::exp(-328.938629f);` is being called with the weights of your true model. And this results in `NaN`.

At least my g++ version (`5.4.0`) returns `NaN` for all values smaller than roughly 87.

```cpp
#include <cmath>
#include <fenv.h>
#include <iostream>

int main()
{
    feenableexcept(FE_ALL_EXCEPT & ~FE_INEXACT);
    for (float x = 30; x > -1000; --x)
    {
        float y = std::exp(x);
        std::cout << x << "\t\t" << y << std::endl;
    }
}
```

Output:

```
    30              1.06865e+13
    29              3.93133e+12
    28              1.44626e+12
    27              5.32048e+11
    ...
    -84             3.3057e-37
    -85             1.2161e-37
    -86             4.47378e-38
    -87             1.64581e-38
    Floating point exception
```

I've just [asked on SO](https://stackoverflow.com/questions/49467244/stdexp-100-0f-nan) if this is intended. ;)

**Comment by Dobiasd** [comment]
Correction: This is not the source of the `NaN`. The real reason is in layer `activation_4` (softmax) the sum of all outputs
```cpp
                for (size_t z_class = 0; z_class < input.shape().depth_; ++z_class)
                {
                    sum += output.get(z_class, y, x);
                }
```
is `0`.

Thus dividing by it
```cpp
                    output.set(z_class, y, x, output.get(z_class, y, x) / sum);
```
produces the `NaN`.

**Comment by Dobiasd** [comment]
Just [commited something](https://github.com/Dobiasd/frugally-deep/commit/ebfc6ea57c882becc048b72f2fd3eeee6210b112) that avoids the division by zero. Now your `true_model` no longer produces the test error with `NaN`, but this one:

```
true_model.txt
Parsing json ... done. elapsed time: 0.006124 s
Running test 1 of 1 ... done. elapsed time: 0.000294 s
ERROR: test failed: output=0 pos=1,0,0 value=0 target=1
```

So I'm sorry, but currently I have no better solution for you than to call `load_model` with `verify = false` as you already mentioned, or using `double` precision instead of `float` precision (`#define FDEEP_FLOAT_TYPE double`).

However I think there is a problem with the weights in your model. Something that results in the `sum` in a `softmax` layer as being `0` presumably is not very meaningful.
Does your model actually produce good results on your training/validation/test sets?

**Comment by huiyiqun** [comment]
Interesting. The model works correctly.

Thanks for you work!
